In [72]:
import pandas as pd
import numpy as np
import os
from pandas import concat
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objs as go
import cufflinks as cf
from plotly.offline import plot, iplot
import plotly.offline as py
import seaborn as sns
from pycaret.classification import *
from pycaret.utils import check_metric
import sweetviz as sv

In [23]:
cf.go_offline()

#### Fonte: https://www.kaggle.com/c/customer-churn-prediction-2020

In [7]:
df_churn = pd.read_csv('/home/eduardo/customer-churn-prediction-2020/sampleSubmission.csv')

In [8]:
df_churn

id churn
0      1   yes
1      2    no
2      3    no
3      4   yes
4      5   yes
..   ...   ...
745  746   yes
746  747   yes
747  748    no
748  749   yes
749  750   yes

[750 rows x 2 columns]

In [9]:
df_test = pd.read_csv('/home/eduardo/customer-churn-prediction-2020/test.csv')

In [10]:
df_test.drop(['id'], axis=1, inplace=True)

In [11]:
df_test

state  account_length      area_code international_plan voice_mail_plan  \
0      KS             128  area_code_415                 no             yes   
1      AL             118  area_code_510                yes              no   
2      IA              62  area_code_415                 no              no   
3      VT              93  area_code_510                 no              no   
4      NE             174  area_code_415                 no              no   
..    ...             ...            ...                ...             ...   
745    GA             130  area_code_415                 no              no   
746    WA              73  area_code_408                 no              no   
747    WV             152  area_code_415                 no              no   
748    DC              61  area_code_415                 no              no   
749    DC             109  area_code_510                 no              no   

     number_vmail_messages  total_day_minutes  total_day_calls  \
0                       25              265.1              110   
1                        0              223.4               98   
2                        0              120.7               70   
3                        0              190.7              114   
4                        0              124.3               76   
..                     ...                ...              ...   
745                      0              119.4               99   
746                      0              177.2              118   
747                      0              184.2               90   
748                      0              140.6               89   
749                      0              188.8               67   

     total_day_charge  total_eve_minutes  total_eve_calls  total_eve_charge  \
0               45.07              197.4               99             16.78   
1               37.98              220.6              101             18.75   
2               20.52              307.2               76             26.11   
3               32.42              218.2              111             18.55   
4               21.13              277.1              112             23.55   
..                ...                ...              ...               ...   
745             20.30              226.3               97             19.24   
746             30.12              270.5               84             22.99   
747             31.31              256.8               73             21.83   
748             23.90              172.8              128             14.69   
749             32.10              171.7               92             14.59   

     total_night_minutes  total_night_calls  total_night_charge  \
0                  244.7                 91               11.01   
1                  203.9                118                9.18   
2                  203.0                 99                9.14   
3                  129.6                121                5.83   
4                  250.7                115               11.28   
..                   ...                ...                 ...   
745                202.7                111                9.12   
746                241.8                112               10.88   
747                213.6                113                9.61   
748                212.4                 97                9.56   
749                224.4                 89               10.10   

     total_intl_minutes  total_intl_calls  total_intl_charge  \
0                  10.0                 3               2.70   
1                   6.3                 6               1.70   
2                  13.1                 6               3.54   
3                   8.1                 3               2.19   
4                  15.5                 5               4.19   
..                  ...               ...                ...   
745                11.3                 7               3.05

In [13]:
df_t = concat([df_test,df_churn ], axis=1)

In [9]:
df_t.drop(['id'], axis=1, inplace=True)

In [14]:
df_t

state  account_length      area_code international_plan voice_mail_plan  \
0      KS             128  area_code_415                 no             yes   
1      AL             118  area_code_510                yes              no   
2      IA              62  area_code_415                 no              no   
3      VT              93  area_code_510                 no              no   
4      NE             174  area_code_415                 no              no   
..    ...             ...            ...                ...             ...   
745    GA             130  area_code_415                 no              no   
746    WA              73  area_code_408                 no              no   
747    WV             152  area_code_415                 no              no   
748    DC              61  area_code_415                 no              no   
749    DC             109  area_code_510                 no              no   

     number_vmail_messages  total_day_minutes  total_day_calls  \
0                       25              265.1              110   
1                        0              223.4               98   
2                        0              120.7               70   
3                        0              190.7              114   
4                        0              124.3               76   
..                     ...                ...              ...   
745                      0              119.4               99   
746                      0              177.2              118   
747                      0              184.2               90   
748                      0              140.6               89   
749                      0              188.8               67   

     total_day_charge  total_eve_minutes  total_eve_calls  total_eve_charge  \
0               45.07              197.4               99             16.78   
1               37.98              220.6              101             18.75   
2               20.52              307.2               76             26.11   
3               32.42              218.2              111             18.55   
4               21.13              277.1              112             23.55   
..                ...                ...              ...               ...   
745             20.30              226.3               97             19.24   
746             30.12              270.5               84             22.99   
747             31.31              256.8               73             21.83   
748             23.90              172.8              128             14.69   
749             32.10              171.7               92             14.59   

     total_night_minutes  total_night_calls  total_night_charge  \
0                  244.7                 91               11.01   
1                  203.9                118                9.18   
2                  203.0                 99                9.14   
3                  129.6                121                5.83   
4                  250.7                115               11.28   
..                   ...                ...                 ...   
745                202.7                111                9.12   
746                241.8                112               10.88   
747                213.6                113                9.61   
748                212.4                 97                9.56   
749                224.4                 89               10.10   

     total_intl_minutes  total_intl_calls  total_intl_charge  \
0                  10.0                 3               2.70   
1                   6.3                 6               1.70   
2                  13.1                 6               3.54   
3                   8.1                 3               2.19   
4                  15.5                 5               4.19   
..                  ...               ...                ...   
745                11.3                 7               3.05

In [15]:
X_train = pd.read_csv('/home/eduardo/customer-churn-prediction-2020/train.csv')

In [16]:
X_train

state  account_length      area_code international_plan voice_mail_plan  \
0       OH             107  area_code_415                 no             yes   
1       NJ             137  area_code_415                 no              no   
2       OH              84  area_code_408                yes              no   
3       OK              75  area_code_415                yes              no   
4       MA             121  area_code_510                 no             yes   
...    ...             ...            ...                ...             ...   
4245    MT              83  area_code_415                 no              no   
4246    WV              73  area_code_408                 no              no   
4247    NC              75  area_code_408                 no              no   
4248    HI              50  area_code_408                 no             yes   
4249    VT              86  area_code_415                 no             yes   

      number_vmail_messages  total_day_minutes  total_day_calls  \
0                        26              161.6              123   
1                         0              243.4              114   
2                         0              299.4               71   
3                         0              166.7              113   
4                        24              218.2               88   
...                     ...                ...              ...   
4245                      0              188.3               70   
4246                      0              177.9               89   
4247                      0              170.7              101   
4248                     40              235.7              127   
4249                     34              129.4              102   

      total_day_charge  total_eve_minutes  total_eve_calls  total_eve_charge  \
0                27.47              195.5              103             16.62   
1                41.38              121.2              110             10.30   
2                50.90               61.9               88              5.26   
3                28.34              148.3              122             12.61   
4                37.09              348.5              108             29.62   
...                ...                ...              ...               ...   
4245             32.01              243.8               88             20.72   
4246             30.24              131.2               82             11.15   
4247             29.02              193.1              126             16.41   
4248             40.07              223.0              126             18.96   
4249             22.00              267.1              104             22.70   

      total_night_minutes  total_night_calls  total_night_charge  \
0                   254.4                103               11.45   
1                   162.6                104                7.32   
2                   196.9                 89                8.86   
3                   186.9                121                8.41   
4                   212.6                118                9.57   
...                   ...                ...                 ...   
4245                213.7                 79                9.62   
4246                186.2                 89                8.38   
4247                129.1                104                5.81   
4248                297.5                116               13.39   
4249                154.8                100                6.97   

      total_intl_minutes  total_intl_calls  total_intl_charge  \
0                   13.7                 3               3.70   
1                   12.2                 5               3.29   
2                    6.6                 7               1.78   
3                   10.1                 3               2.73   
4                    7.5                 7               2.03   
...                  ...               ...                ...   
4245  

In [17]:
print('train', X_train.shape)
print('test', df_test.shape)

train (4250, 20)
test (750, 19)


In [18]:
X_train.isnull().sum()

state                            0
account_length                   0
area_code                        0
international_plan               0
voice_mail_plan                  0
number_vmail_messages            0
total_day_minutes                0
total_day_calls                  0
total_day_charge                 0
total_eve_minutes                0
total_eve_calls                  0
total_eve_charge                 0
total_night_minutes              0
total_night_calls                0
total_night_charge               0
total_intl_minutes               0
total_intl_calls                 0
total_intl_charge                0
number_customer_service_calls    0
churn                            0
dtype: int64

In [19]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4250 entries, 0 to 4249
Data columns (total 20 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   state                          4250 non-null   object 
 1   account_length                 4250 non-null   int64  
 2   area_code                      4250 non-null   object 
 3   international_plan             4250 non-null   object 
 4   voice_mail_plan                4250 non-null   object 
 5   number_vmail_messages          4250 non-null   int64  
 6   total_day_minutes              4250 non-null   float64
 7   total_day_calls                4250 non-null   int64  
 8   total_day_charge               4250 non-null   float64
 9   total_eve_minutes              4250 non-null   float64
 10  total_eve_calls                4250 non-null   int64  
 11  total_eve_charge               4250 non-null   float64
 12  total_night_minutes            4250 non-null   f

Visualizacao dos dados com sweetviz.

In [83]:
my_report = sv.analyze(X_train, 'churn')
my_report.show_html()

                                             |      | [  0%]   00:00 -> (? left)

Report SWEETVIZ_REPORT.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


In [63]:
#YES CHURN X NO CHURN
X_train.churn.value_counts().iplot(kind='bar', color='red')

### Analisando a correlacao variaveis x churn

In [81]:
#sns.barplot('churn', 'total_day_minutes', data = X_train)
#TOTAL_DAY_MINUTES X CHURN
fig = px.histogram(data_frame=X_train, x="total_day_minutes", color="churn", width=900, height=400)
fig.show()

In [84]:
#Total_Day_Charge x Churn
fig = px.histogram(data_frame=X_train, x="total_day_charge", color="churn", width=900, height=400)
fig.show()

In [85]:
#Number_Vmail_Messages x Churn
fig = px.histogram(data_frame=X_train, x="number_vmail_messages", color="churn", width=900, height=400)
fig.show()

In [87]:
#Total_Eve_Minutes x Churn
fig = px.histogram(data_frame=X_train, x="total_eve_minutes", color="churn", width=900, height=400)
fig.show()

In [89]:
#Account_Length x Churn
fig = px.histogram(data_frame=X_train, x="account_length", color="churn", width=900, height=400)
fig.show()

In [66]:
X_train.state.unique()

array(['OH', 'NJ', 'OK', 'MA', 'MO', 'LA', 'WV', 'IN', 'RI', 'IA', 'MT',
       'NY', 'ID', 'VA', 'TX', 'FL', 'CO', 'AZ', 'SC', 'WY', 'HI', 'NH',
       'AK', 'GA', 'MD', 'AR', 'WI', 'OR', 'MI', 'DE', 'UT', 'CA', 'SD',
       'NC', 'WA', 'MN', 'NM', 'NV', 'DC', 'VT', 'KY', 'ME', 'MS', 'AL',
       'NE', 'KS', 'TN', 'IL', 'PA', 'CT', 'ND'], dtype=object)

In [61]:
#Estados com o maior numero de clientes
X_train.state.value_counts().iplot(kind='bar', color='lightgreen')

In [82]:
#State x Churn
fig = px.histogram(data_frame=X_train, x="state", color="churn", width=900, height=400)
fig.show()

In [88]:
#Area_Code x Churn
fig = px.histogram(data_frame=X_train, x="area_code", color="churn", width=900, height=400)
fig.show()

# Principais fatores de rotatividade
* A taxa de rotatividade é alta quando O total_day_minutes está entre 210 e 300 min.
* Quando o total_day_charge (cobranca de ligacoes diurnas) está entre 40 e 50.
* Quando nao tem o servico de number_vmail_messages.
* Quando o account_length (numero de meses que o cliente esta utilizando o provedor) esta entre 60 e 120.
* Quando o total_day_calls (total de ligacoes por dia)esta entre 85 e 115.
* Quando o total_eve_minutes está entre 160 e 260 minutos.
* Quando o total_eve_calls esta entre 90 e 115.
* Quando o total_eve_charge esta entre 15 e 18.
* Quando o total_night_minutes esta entre 190 e 220 minutos.
* Quando o total_nigh_calls esta entre 90 e 110.
* Quando o total_night_charge esta entre 7.5 e 10.
* Quando o total_intl_minutes esta entre 9 e 12.
* Quando o total_intl_calls é 1.
* Quando o total_intl_charge esta entre 2.5 e 3.
* Quando o area_code é = '415' (Canada).

### Pre-processamento dos dados com pycaret

In [18]:
exp_clf01 = setup(data = X_train
                  ,target = 'churn')

IntProgress(value=0, description='Processing: ', max=3)

Initiated  . . . . . . . . . . . . . . . . . .            16:28:29
Status     . . . . . . . . . . . . . . . . . .  Preprocessing Data

Text(value="Following data types have been inferred automatically, if they are correct press enter to continue…

Data Type
state                          Categorical
account_length                     Numeric
area_code                      Categorical
international_plan             Categorical
voice_mail_plan                Categorical
number_vmail_messages              Numeric
total_day_minutes                  Numeric
total_day_calls                    Numeric
total_day_charge                   Numeric
total_eve_minutes                  Numeric
total_eve_calls                    Numeric
total_eve_charge                   Numeric
total_night_minutes                Numeric
total_night_calls                  Numeric
total_night_charge                 Numeric
total_intl_minutes                 Numeric
total_intl_calls                   Numeric
total_intl_charge                  Numeric
number_customer_service_calls  Categorical
churn                                Label

quit


SystemExit: Read the documentation of setup to learn how to overwrite data types over the inferred types. setup function must run again before you continue modeling.

In [15]:
exp_clf01

(      account_length  number_vmail_messages  total_day_minutes  \
 1766            82.0                    0.0         154.000000   
 3639           105.0                    0.0         274.000000   
 3930            81.0                    0.0          79.400002   
 942             91.0                    0.0         154.399994   
 2301           189.0                   30.0         155.199997   
 ...              ...                    ...                ...   
 3829           126.0                   24.0         197.300003   
 2366           169.0                    0.0         142.500000   
 212             92.0                    0.0          62.599998   
 3712           130.0                    0.0         143.800003   
 3198            81.0                    0.0         139.800003   
 
       total_day_calls  total_eve_calls  total_eve_charge  total_night_minutes  \
 1766            107.0            114.0          8.020000           287.600006   
 3639            156.0        

In [16]:
df_transformed = get_config('X_train')

In [17]:
df_transformed

account_length  number_vmail_messages  total_day_minutes  \
152             73.0                    0.0         214.300003   
2768           131.0                    0.0         175.100006   
3264            49.0                   42.0         159.000000   
590             83.0                   32.0          94.699997   
459             58.0                   20.0         194.500000   
...              ...                    ...                ...   
2909           103.0                    0.0         147.300003   
3935            96.0                    0.0         166.500000   
2088            45.0                    0.0          96.099998   
1395            84.0                   12.0          89.699997   
2667            35.0                    0.0         138.100006   

      total_day_calls  total_eve_calls  total_eve_charge  total_night_minutes  \
152             145.0            135.0             22.82           241.199997   
2768             73.0            116.0             14.61           131.100006   
3264             67.0            125.0             10.74           206.800003   
590             111.0             98.0             13.12           200.399994   
459             110.0             89.0             18.16           236.600006   
...               ...              ...               ...                  ...   
2909             81.0             88.0             13.46           193.800003   
3935            101.0            108.0              6.66           221.899994   
2088            103.0            134.0             20.98           229.699997   
1395             87.0             73.0             11.78           165.800003   
2667            115.0             82.0             13.45           215.699997   

      total_night_calls  total_intl_charge  state_AK  ...  \
152                92.0               2.92       0.0  ...   
2768               94.0               1.97       0.0  ...   
3264               88.0               2.00       0.0  ...   
590               109.0               2.86       0.0  ...   
459                92.0               2.57       0.0  ...   
...                 ...                ...       ...  ...   
2909              103.0               2.97       0.0  ...   
3935              112.0               3.11       0.0  ...   
2088               92.0               2.62       0.0  ...   
1395              114.0               2.89       0.0  ...   
2667              118.0               2.78       0.0  ...   

      number_customer_service_calls_0  number_customer_service_calls_1  \
152                               0.0                              1.0   
2768                              0.0                              1.0   
3264                              0.0                              0.0   
590                               0.0                              0.0   
459                               0.0                              1.0   
...                               ...                              ...   
2909                              0.0                              0.0   
3935                              0.0                              0.0   
2088                              0.0                              1.0   
1395                              0.0                              1.0   
2667                              0.0                              0.0   

      number_customer_service_calls_2  number_customer_service_calls_3  \
152                               0.0                              0.0   
2768                              0.0                              0.0   
3264                              1.0                              0.0   
590                               1.0                              0.0   
459                               0.0                              0.0   
...                               ...                              ...   
2909                              1.0                              0.0   
3935                      

### Comparando os modelos.

In [18]:
best_model = compare_models()

In [19]:
print(best_model)

LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               importance_type='split', learning_rate=0.1, max_depth=-1,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
               random_state=2439, reg_alpha=0.0, reg_lambda=0.0, silent=True,
               subsample=1.0, subsample_for_bin=200000, subsample_freq=0)


### Criando o modelo.

In [20]:
xgbc = create_model('xgboost')

### Tunando o modelo.

In [21]:
tuned_xgbc = tune_model(xgbc)

### Personalizando a busca por hiperparametros.

In [22]:
params = {
    'n_estimators': np.arange(100,500,50),
    'max_depth': [-1,3,5,10],
    'num_leaves': np.arange(50,120,10)
}

In [23]:
params

{'n_estimators': array([100, 150, 200, 250, 300, 350, 400, 450]),
 'max_depth': [-1, 3, 5, 10],
 'num_leaves': array([ 50,  60,  70,  80,  90, 100, 110])}

In [24]:
tuned_xgbc2 = tune_model(xgbc, custom_grid = params)

In [25]:
print(tuned_xgbc2)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, enable_categorical=False,
              gamma=0, gpu_id=-1, importance_type=None,
              interaction_constraints='', learning_rate=0.300000012,
              max_delta_step=0, max_depth=5, min_child_weight=1, missing=nan,
              monotone_constraints='()', n_estimators=100, n_jobs=-1,
              num_leaves=80, num_parallel_tree=1, objective='binary:logistic',
              predictor='auto', random_state=2439, reg_alpha=0, reg_lambda=1,
              scale_pos_weight=1, subsample=1, tree_method='auto',
              use_label_encoder=True, validate_parameters=1, ...)


### Analisando a performance do modelo.

In [26]:
evaluate_model(tuned_xgbc2)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Hyperparameters', 'param…

### Predição utilizando a amostra separada pra teste.

In [39]:
predict_model(tuned_xgbc2)

account_length  number_vmail_messages  total_day_minutes  \
0               82.0                    0.0         154.000000   
1              105.0                    0.0         274.000000   
2               81.0                    0.0          79.400002   
3               91.0                    0.0         154.399994   
4              189.0                   30.0         155.199997   
...              ...                    ...                ...   
1271           126.0                   24.0         197.300003   
1272           169.0                    0.0         142.500000   
1273            92.0                    0.0          62.599998   
1274           130.0                    0.0         143.800003   
1275            81.0                    0.0         139.800003   

      total_day_calls  total_eve_calls  total_eve_charge  total_night_minutes  \
0               107.0            114.0          8.020000           287.600006   
1               156.0             69.0         22.360001           195.899994   
2                99.0             90.0         12.300000           204.699997   
3               165.0            121.0         14.310000           239.899994   
4               116.0             50.0         16.620001           170.100006   
...               ...              ...               ...                  ...   
1271             82.0             51.0         14.250000           222.000000   
1272             82.0            110.0         19.670000           131.199997   
1273            111.0            126.0         15.350000           221.699997   
1274             76.0            103.0         23.360001           194.699997   
1275             98.0            103.0         21.010000           241.000000   

      total_night_calls  total_intl_charge  state_AK  ...  \
0                  95.0               2.73       0.0  ...   
1                  93.0               3.05       0.0  ...   
2                 114.0               2.46       0.0  ...   
3                  81.0               3.16       0.0  ...   
4                 108.0               4.16       0.0  ...   
...                 ...                ...       ...  ...   
1271               89.0               2.00       0.0  ...   
1272               67.0               2.70       0.0  ...   
1273               80.0               2.81       0.0  ...   
1274              105.0               3.32       0.0  ...   
1275              113.0               3.16       0.0  ...   

      number_customer_service_calls_3  number_customer_service_calls_4  \
0                                 0.0                              0.0   
1                                 1.0                              0.0   
2                                 0.0                              0.0   
3                                 0.0                              0.0   
4                                 0.0                              0.0   
...                               ...                              ...   
1271                              0.0                              0.0   
1272                              0.0                              0.0   
1273                              0.0                              0.0   
1274                              0.0                              0.0   
1275                              1.0                              0.0   

      number_customer_service_calls_5  number_customer_service_calls_6  \
0                                 0.0                              0.0   
1                                 0.0                              0.0   
2                                 0.0                              0.0   
3                                 1.0                              0.0   
4                                 0.0                              0.0   
...                               ...                              ...   
1271                              0.0                              0.0   
1272                      

### Cross Validation

In [51]:
df = validation_predictions = predict_model(tuned_xgbc2, data = df_t)
validation_predictions.head()

state  account_length      area_code international_plan voice_mail_plan  \
0    KS             128  area_code_415                 no             yes   
1    AL             118  area_code_510                yes              no   
2    IA              62  area_code_415                 no              no   
3    VT              93  area_code_510                 no              no   
4    NE             174  area_code_415                 no              no   

   number_vmail_messages  total_day_minutes  total_day_calls  \
0                     25              265.1              110   
1                      0              223.4               98   
2                      0              120.7               70   
3                      0              190.7              114   
4                      0              124.3               76   

   total_day_charge  total_eve_minutes  ...  total_night_minutes  \
0             45.07              197.4  ...                244.7   
1             37.98              220.6  ...                203.9   
2             20.52              307.2  ...                203.0   
3             32.42              218.2  ...                129.6   
4             21.13              277.1  ...                250.7   

   total_night_calls  total_night_charge  total_intl_minutes  \
0                 91               11.01                10.0   
1                118                9.18                 6.3   
2                 99                9.14                13.1   
3                121                5.83                 8.1   
4                115               11.28                15.5   

   total_intl_calls  total_intl_charge  number_customer_service_calls  churn  \
0                 3               2.70                              1    yes   
1                 6               1.70                              0     no   
2                 6               3.54                              4     no   
3                 3               2.19                              3    yes   
4                 5               4.19                              3    yes   

   Label   Score  
0     no  0.9927  
1     no  0.9071  
2    yes  0.8864  
3     no  0.9980  
4     no  0.9975  

[5 rows x 22 columns]

### Finalize Model

In [36]:
final_xgbc = finalize_model(tuned_xgbc2)

In [32]:
#save_model(final_xgbc,'Final XGBC Model')

Transformation Pipeline and Model Successfully Saved


In [34]:
#final_xgbc.to_csv('XGBC', sep=',', index=False)

In [52]:
df.to_csv('churn_predict.csv', sep=',', index=False)

In [53]:
!ls

 churn_predict.csv	 logs.log	        test.csv
 churn_pred.ipynb	 sampleSubmission.csv   train.csv
'Final XGBC Model.pkl'	 SWEETVIZ_REPORT.html   Untitled.ipynb
